### PR TITLE
Add DelegatingSecurityContextTaskScheduler

### PR DESCRIPTION
**Pull request by richardvaldiviesomacias**
Spring Security provides a TaskScheduler variant or the documentation routes users to expose a TaskScheduler bean rather than exposing a ScheduledExecutorService bean.

New file: DelegatingSecurityContextTaskScheduler which is a DelegatingSecurityContextAsyncTaskExecutor thant implements TaskScheduler.

ExplicitSecurityContextSchedulingTaskSchedulerTests and AbstractSecurityContextTaskSchedulerTests for testing.

### CODE REVIEW

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		this.taskScheduler = taskScheduler;
+	}
+
+	public DelegatingSecurityContextTaskScheduler(
+			AsyncTaskExecutor delegateAsyncTaskExecutor, TaskScheduler taskScheduler) {
+		super(delegateAsyncTaskExecutor);
+		this.taskScheduler = taskScheduler;
+	}
+
+	private SchedulingTaskExecutor getDelegate() {
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: Is this necessary?

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * An {@link SchedulingTaskExecutor} which wraps each {@link Runnable} in a
+ * {@link DelegatingSecurityContextRunnable} and each {@link Callable} in a
+ * {@link DelegatingSecurityContextCallable}.
+ *
+ * @author Richard Valdivieso
+ * @since 5.1
+ *
+ */
+public class DelegatingSecurityContextTaskScheduler extends
+		DelegatingSecurityContextAsyncTaskExecutor implements TaskScheduler {
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: We should not extend `DelegatingSecurityContextAsyncTaskExecutor` since `TaskExecutor` is not a `TaskExecutor`. Instead the class can use it's own wrap method directly.

File: core/src/test/java/org/springframework/security/scheduling/AbstractSecurityContextTaskSchedulerTests.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * An {@link SchedulingTaskExecutor} which wraps each {@link Runnable} in a
+ * {@link DelegatingSecurityContextRunnable} and each {@link Callable} in a
+ * {@link DelegatingSecurityContextCallable}.
+ *
+ * @author Richard Valdivieso
+ * @since 5.1
+ * @see ExplicitSecurityContextSchedulingTaskSchedulerTests
+ *
+ */
+public abstract class AbstractSecurityContextTaskSchedulerTests extends
```
Comment (rwinch) [inline core/src/test/java/org/springframework/security/scheduling/AbstractSecurityContextTaskSchedulerTests.java]: It's nice to add tests, but we should do this in a separate PR

File: core/src/test/java/org/springframework/security/scheduling/ExplicitSecurityContextSchedulingTaskSchedulerTests.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Tests Explicitly specifying the {@link SecurityContext} on
+ * {@link DelegatingSecurityContextTaskScheduler}
+ *
+ * @author Richard Valdivieso
+ * @since 5.1
+ *
+ */
+public class ExplicitSecurityContextSchedulingTaskSchedulerTests extends
```
Comment (rwinch) [inline core/src/test/java/org/springframework/security/scheduling/ExplicitSecurityContextSchedulingTaskSchedulerTests.java]: It's nice to add tests, but we should do this in a separate PR

**Comment by richardvaldiviesomacias** [comment]
I will make those changes.

File: core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Test An implementation of {@link TaskScheduler}  invoking it whenever the trigger
+ * indicates a next execution time.
+ *
+ * @author Richard Valdivieso
+ * @since 5.1
+ *
+ */
+public class DelegatingSecurityContextTaskSchedulerTest {
```
Comment (rwinch) [inline core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTest.java]: Spring ends tests with `Tests` rather than `Test`. Can you please rename to `DelegatingSecurityContextTaskSchedulerTests`?

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public class DelegatingSecurityContextTaskScheduler implements TaskScheduler {
+
+	private final TaskScheduler taskScheduler;
+
+	/**
+	 * Creates a new {@link DelegatingSecurityContextTaskScheduler}
+	 * @param taskScheduler the {@link TaskScheduler}
+	 */
+
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: Remove new line

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	public DelegatingSecurityContextTaskScheduler(TaskScheduler taskScheduler) {
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: Add an `Assert.notNull` and a test for it

Comment (richardvaldiviesomacias) [reply core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: fix

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java (hunk 6 above)
Comment (richardvaldiviesomacias) [reply core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: fix

File: core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTest.java (hunk 5 above)
Comment (richardvaldiviesomacias) [reply core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTest.java]: done

**Comment by richardvaldiviesomacias** [comment]
Done. Thanks for all your help 😁

**Comment by rwinch** [comment]
@richardvaldiviesomacias Thanks for the fast turnaround! This is now merged into master. Thank you for taking the time to make Spring better. We hope to see you again soon (submitting a PR, issue, etc)! If you are looking for issues to help with the label [Help Wanted](https://github.com/spring-projects/spring-security/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22help+wanted%22) is a sign that the issue is relatively small. Just comment on the issue to claim it before starting work.
